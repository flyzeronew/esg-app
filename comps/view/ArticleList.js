import styles from "./ArticleList.module.css";

export const Article = ({ articleData, index }) => {
    return (
        <>
            <li key={index}>
                <a href={`/view/${articleData.article_genres[0].en_name}/${articleData.id}`}>
                    <div className={styles.img}>
                        <img src={articleData.cover_img ? articleData.cover_img : process.env.IMG_DEFAULT} alt="img" width={559} height={315} loading="lazy" />
                    </div>
                    <h2 className={styles.txt}>{articleData.title}</h2>
                    <div className={styles.name}>
                        {articleData.partner != null ?
                            <>
                                <div className={styles.nameImg}>
                                    <div className={styles.img1} >
                                        <img src={articleData.partner.avatar} alt="logo" width={50} height={50} loading="lazy" />
                                    </div>
                                </div>
                                <p>{articleData.partner.name}</p>
                            </>
                            : <p>{articleData.author_name}</p>}
                    </div>
                </a>
            </li>
    
        </>
    )
}
const ArticleList = ({ articleList }) => {
    return (
        <div className={styles.list}>
            <ul>
                {articleList?.map((articleData, index) => (
                    <Article key={index} articleData={articleData} index={index}></Article>
                ))
                }
            </ul>
        </div>
    )
}


export default ArticleList;