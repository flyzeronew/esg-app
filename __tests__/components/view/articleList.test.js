import { render, screen, logRoles } from '@testing-library/react';
import ArticleList, { Article } from '@/comps/view/ArticleList';

process.env.IMG_DEFAULT = '/default-esg-image.jpg';

describe("Article Component",()=>{
    const mockAtlWithPartner = {
        id: '1',
        article_genres: [{ en_name: 'environment' }],
        cover_img: 'https://esg.testing.tvbs.com.tw/test.jpg',
        title: 'Partner title here',
        partner: {
          avatar: 'https://esg.testing.tvbs.com.tw/testavt.jpg',
          name: 'Partner test'
        }
      };
    
      const mockAtlWithoutPartner = {
        id: '2',
        article_genres: [{ en_name: 'environment' }],
        cover_img: null,
        title: 'Another Test Article',
        author_name: 'Test Name',
        partner: null
      };
    
    it("renders article unchanged",()=>{
        const {container}  = render(<Article articleData={mockAtlWithPartner} index={1}></Article>)
        expect(container).toMatchSnapshot();
    })
     it("renders article with partner data",()=>{
         render(<Article articleData={mockAtlWithPartner} index={1}></Article>)

        //link check
        expect(screen.getByRole('link')).toHaveAttribute("href", "/view/environment/1");

        //title
        expect(screen.getByText(mockAtlWithPartner.title)).toBeInTheDocument();
        expect(screen.getAllByAltText("img")[0]).toHaveAttribute("src", mockAtlWithPartner.cover_img)
        expect(screen.getByText(mockAtlWithPartner.partner.name)).toBeInTheDocument();

     })
     it("renders article without partner data",()=>{
        render(<Article articleData={mockAtlWithoutPartner} index={1}></Article>)
        expect(screen.getByText(mockAtlWithoutPartner.author_name)).toBeInTheDocument();
        expect(screen.getAllByAltText("img")[0]).toHaveAttribute("src",process.env.IMG_DEFAULT)
     })
})

describe("Article List component",()=>{
    const mockArticleListData = [
        {
            id: '1',
            article_genres: [{ en_name: 'environment' }],
            cover_img: 'https://esg.testing.tvbs.com.tw/test.jpg',
            title: 'Partner title here',
            partner: {
              avatar: 'https://esg.testing.tvbs.com.tw/testavt.jpg',
              name: 'Partner test'
            }
          },
          {
            id: '2',
            article_genres: [{ en_name: 'environment' }],
            cover_img: 'https://esg.testing.tvbs.com.tw/test.jpg',
            title: 'Another Test Article',
            author_name: 'Test Name',
            partner: null
          }
      ];
    it("renders the Article list unchanged",()=>{
        const { container } = render(<ArticleList articleList={mockArticleListData}></ArticleList>);
        expect(container).toMatchSnapshot();
    })

    it("renders the list of articles provided",()=>{
       const {container} =  render(<ArticleList articleList={mockArticleListData}></ArticleList>);
        // logRoles(container);
        //check for number of list items
        expect(screen.getAllByRole("listitem")).toHaveLength(2);

    })
    it("renders the zero article",()=>{
        render(<ArticleList articleList={[]}></ArticleList>);
        expect(screen.queryAllByRole("listitem")).toHaveLength(0);
    })
})
