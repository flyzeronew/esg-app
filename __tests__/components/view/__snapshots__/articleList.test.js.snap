// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Component renders article unchanged 1`] = `
<div>
  <li>
    <a
      href="/view/environment/1"
    >
      <div
        class="img"
      >
        <img
          alt="img"
          height="315"
          loading="lazy"
          src="https://esg.testing.tvbs.com.tw/test.jpg"
          width="559"
        />
      </div>
      <h2
        class="txt"
      >
        Partner title here
      </h2>
      <div
        class="name"
      >
        <div
          class="nameImg"
        >
          <div
            class="img1"
          >
            <img
              alt="logo"
              height="50"
              loading="lazy"
              src="https://esg.testing.tvbs.com.tw/testavt.jpg"
              width="50"
            />
          </div>
        </div>
        <p>
          Partner test
        </p>
      </div>
    </a>
  </li>
</div>
`;

exports[`Article List component renders the Article list unchanged 1`] = `
<div>
  <div
    class="list"
  >
    <ul>
      <li>
        <a
          href="/view/environment/1"
        >
          <div
            class="img"
          >
            <img
              alt="img"
              height="315"
              loading="lazy"
              src="https://esg.testing.tvbs.com.tw/test.jpg"
              width="559"
            />
          </div>
          <h2
            class="txt"
          >
            Partner title here
          </h2>
          <div
            class="name"
          >
            <div
              class="nameImg"
            >
              <div
                class="img1"
              >
                <img
                  alt="logo"
                  height="50"
                  loading="lazy"
                  src="https://esg.testing.tvbs.com.tw/testavt.jpg"
                  width="50"
                />
              </div>
            </div>
            <p>
              Partner test
            </p>
          </div>
        </a>
      </li>
      <li>
        <a
          href="/view/environment/2"
        >
          <div
            class="img"
          >
            <img
              alt="img"
              height="315"
              loading="lazy"
              src="https://esg.testing.tvbs.com.tw/test.jpg"
              width="559"
            />
          </div>
          <h2
            class="txt"
          >
            Another Test Article
          </h2>
          <div
            class="name"
          >
            <p>
              Test Name
            </p>
          </div>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
